odoo/websocket.py

```diff
@@ -0,0 +1,883 @@
+        if ctx is not None:
+            self.request.update_env(context=ctx)
+        result = self.request.registry['ir.websocket']._dispatch(endpoint)
+        return self._response(result)
+
+    def post_dispatch(self, response):
+        pass
+
+    def handle_error(self, exc):
+        return self._response(error=exc)
+
+    @classmethod
+    def is_compatible_with(cls, request):
+        return isinstance(request, WebsocketRequest)
+
+    def _response(self, response=None, error=None):
+        if error:
+            return {
+                'error': self._serialize_exception(
+                    error, self.request.client_uuid
+                )
+            }
+        return response
+
+    def _serialize_exception(self, exc, client_uuid):
+        serialized_exc = http.serialize_exception(exc)
+        serialized_exc['client_uuid'] = client_uuid
+        return serialized_exc
+
+
+class WebsocketRequest:
+    def __init__(self, websocket, session, db, httprequest, data):
+        self.data = data
+        self.db = db
+        self.httprequest = httprequest
+        self.session = session
+        self.ws = websocket
+        self.dispatcher = http._dispatchers['websocket'](self)
+        self.client_uuid = None
+        self.path = None
+        self.params = None
+
+    def __enter__(self):
+        _ws_request_stack.push(self)
+        return self
+
+    def __exit__(self, *args):
+        self.ws = None
+        _ws_request_stack.pop()
+
+    def decode_websocket_message(self):
+        try:
+            jsonrequest = json.loads(self.data)
+            self.path = jsonrequest.get('path')
+            self.params = jsonrequest.get('kwargs', {})
+            self.client_uuid = jsonrequest.get('client_uuid')
+            self.session_sid = jsonrequest.get('session_sid')
+            if None in (self.path, self.session_sid):
+                missing_keys = []
+                if not self.path:
+                    missing_keys.append('path')
+                if not self.session_sid:
+                    missing_keys.append('session_sid')
+                raise InvalidWebsocketRequest(
+                    f'{", ".join(missing_keys)} key(s) missing from request'
+                )
+        except ValueError:
+            raise InvalidWebsocketRequest('Invalid JSON data')
+
+    def serve_websocket_message(self):
+        try:
+            self.registry = Registry(self.db)
+            self.registry.check_signaling()
+        except (
+            AttributeError, psycopg2.OperationalError, psycopg2.ProgrammingError
+        ) as exc:
+            raise InvalidDatabaseException() from exc
+
+        with closing(self.registry.cursor()) as cr:
+            self.env = api.Environment(cr, self.session.uid, self.session.context)
+            threading.current_thread().uid = self.env.uid
+            try:
+                return service_model.retrying(self._serve_ir_websocket, self.env)
```

Comment:
The session is not refresh, there is missing code next to https://github.com/odoo/odoo/blob/64745b506c877a493057cc5a502bf38805079d56/odoo/service/model.py#L140-L146